### DISCUSSION TITLE
Point recalculation for a custom shape

### DISCUSSION THREAD

**Discussion by SupremeDeity** [fabricjs/fabric.js#10124] [Q&A] [answered]
Hello. I am currently working on "history" functionality and realized the *points* i pass to my object wouldn't get modified. The width and height of my custom shapes depend on these points, since i want basically a drawing tool of sorts. There are two sets of points:
1) The first set is the starting position where the mouse button was held at to start drawing
2) The second set is the current position of the mouse.

I simply do some calculations based on this. For example for a rectangle:
```js
const [x1, y1, x2, y2] = this.points;
let width = this.roughOptions.size?.width ?? Math.abs(x2 - x1);
let height = this.roughOptions.size?.height ?? Math.abs(y2 - y1);
const left = Math.min(x1, x2);
const top = Math.min(y1, y2);

this.set({
     left: left + width / 2,
    top: top + height / 2,
    width: width,
    height: height,
});
```
![image](https://github.com/user-attachments/assets/85179126-99d6-4b89-b268-923d8fef6abd)

Is there some way i can easily transform this? some hints? I know about `Point.transform(matrix)` but i cant really understand how it would help me. There is no issue moving this object and everything else works fine but its a issue when i try and work on serialization/deserialization as the modified position will never be accounted for due to how the calculations are done.

If more code or anything is needed please tell me and i will try my best to provide that.

**Comment by asturur** [comment]
Is unclear what transformation you need and for what.
Can you specify that?
What is the 'this' in your case?
I see you are using rough.js to do some drawing lines, but all the other context is missing.

**Reply by SupremeDeity** [reply]
`this` is a custom object/shape. I am subclassing from FabricObject. Basically whenever i manually change a property, i call that function which re-calculates the width/height, top and left and regenerates my rough object. Issue is, the points never change after the "drawing" phase is complete, so while fabric.js recalculates the left and top property if i move my object on the canvas, if i ever end up changing a property manually like the stroke or something, those outdated points are going to reset the position because the position (width,height, top and left) depend on the points. Is there some way i can keep track of these transformations so i can more easily recalculate this stuff? I have seen your post on recalculations for the `Path` and `Polygon` but they dont really apply to shapes like these since they dont have a `pathOffset`

**Reply by asturur** [reply]
This seems a basic programming question more than an issue with fabric.
Can you have the constructor do position and size, but have the update function do only the size and don't touch the position anymore?

You have to do something similar as fabricJS does when you update a property that influence the size, otherwise the object will seem to move.

If you are working on a new project please use originX and originY set to 'center' it makes your calculation easier and is better for future updates

**Reply by SupremeDeity** [reply]
@asturur ok i had written quite a lot but i deleted that. I think i have figured out much of the idea now after pulling up the drawing board and mapping things out.

I will post the basic code here later just in case anyone is looking for it, maybe with even a diagram. The basic idea was just to calculate the offset from the starting position and never changing left and top manually (unless in the constructor)

One last thing is, you told to use originX and originY as center to make calculations easy but isnt setting them to 0,0 more convenient + good mechanic? especially for apps that require drawing by actually dragging the mouse. Like diagraming apps like excallidraw. With center, center it will look like the shape(like rectangle) is originating from the center instead of top left like needed. 

Is there any solution to that? Sorry for this i know you're busy but i'm a bit of a beginner so i think im missing some context here.

**Reply by asturur** [reply, marked as answer]
The issue with originX and originY is that they will be removed and the code will behave as if they are in the center.
So if you adapt to center now you will not have issue later.

Is it true that top,left or 0,0 make your life easier when drawing, but that means that you pay the calculation price every single frame.
object.setPositionByOrigin will stay available, so every time you draw the object while drawing if you want to keep the position steady from the perspective of the 0,0 corner or the top,left corner you will grab the correct point when you start your drag and call `setPositionByOrigin(position, 'left', 'top')` every frame, as fabricJS does every single time when you scale or rotate something.

Look inside fabric for the utility called `wrapWithFixedAnchor`


```ts
      const centerPoint = target.getRelativeCenterPoint(), // get the center
      constraint = target.translateToOriginPoint(centerPoint, 'left', 'top'), // find out where top left is
      actionPerformed = actionHandler(eventData, transform, x, y); // do whatever you want to do that change the shape
    target.setPositionByOrigin(
      constraint,
      'left',
      'top',
    ); // put the shape back in the same point

```

**Reply by SupremeDeity** [reply]
Thank you so much!!

**Comment by SupremeDeity** [comment]
Here's the code if anyone is looking for it. This is for useful for cases where you want to implement drawing apps and not make it look like the shape is growing from the center due to origin being set to center,center, as that property is going to be removed later. 
```js
import * as fabric from "fabric";

export class CustomShape extends fabric.Rect {
    static get type() {
        return 'CustomShape';
    }
    constructor(options = {}) {
        super(options);
        // This line might need tweaking in that ternary condition. MAKE SURE TO SET THIS
        this.left = this.left !== 0 ? this.left : options.points[0];
        this.top = this.top !== 0 ? this.top : options.points[1];
        this._updateShape();
    }
    
    // THIS FUNCTION NEEDS TO BE CALLED EVERYTIME SHAPE CHANGES, ALSO EVERY FRAME DURING DRAWING.
    // SIMPLY CHANGE THE x2, y2 POINTS with current mouse position AND KEEP RECALLING IT WHILE in "DRAWING" phase, i.e while mouse button is pressed
    _updateShape() {
        const [x1, y1, x2, y2] = this.points;
        
        // VERY IMPORTANT IF YOUR SHAPE CAN BE MOVED AROUND AND IF YOU IMPLEMENT
        // SERIALIZATION
        const widthOffset = this.left === x1 ? 0 : this.left - x1;
        const heightOffset = this.top === y1 ? 0 : this.top - y1;
        let width = Math.max(Math.abs(x2 - this.left + widthOffset), this.minSize);
        let height = Math.max(Math.abs(y2 - this.top + heightOffset), this.minSize);

        // Gets the top and left based on set origin
        const relativeCenter = this.getRelativeCenterPoint()
        // Translates the relativeCenter point as if origin = 0,0
        const constraint = this.translateToOriginPoint(relativeCenter, 'left', 'top')

        // Shape changing stuff
        this.set({
            width: width,
            height: height,
        });

        // Put shape back in place
        this.setPositionByOrigin(
            constraint,
            'left',
            'top',
        );

        this.setCoords();
    }
```

**Reply by SupremeDeity** [reply]
here points:
1) x1, y1 -> The position where mouse button was pressed down on
2) x2, y2 -> The position changes *while* drawing, its the current mouse position *while* mouse button is still pressed down and moving.

**Reply by SupremeDeity** [reply]
![image](https://github.com/user-attachments/assets/36e4ac5d-ce02-498f-b36a-a44d5bab64c7)
